File: src/main/java/com/oycm/datastructure/binary_tree/top_down_traverse/Solution_10.java
package com.oycm.datastructure.binary_tree.top_down_traverse;

import com.oycm.TreeNode;

public class Solution_10 {

    /**
     * 1022. <a href="https://leetcode.cn/problems/sum-of-root-to-leaf-binary-numbers/description/">从根到叶的二进制数之和</a> 1462
     *
     * @param root
     * @return
     */
    public int sumRootToLeaf(TreeNode root) {
        /*
        和 10 进制路径表示数字类似, ans += 叶子节点表示的数字
         */
        dfs(root, 0);
        return ans;
    }

    int ans = 0;

    public void dfs(TreeNode node, int k) {
        if (node == null) {
            return;
        }
        k = (k << 1) + node.val;
        if (node.left == null && node.right == null) {
            ans += k;
        }
        dfs(node.left, k);
        dfs(node.right, k);
    }


}

File: src/main/java/com/oycm/month2026/september/Solution_10.py
class Solution_10:
    def averageOfSubtree(self, root: TreeNode) -> int:
        # 自底向上返回
        ans = 0

        def dfs(node: TreeNode | Node) -> tuple[int, int]:
            if node is None:
                return 0, 0
            left_sum, left_size = dfs(node.left)
            right_sum, right_size = dfs(node.right)
            sum_ = left_sum + right_sum + node.val
            size = left_size + right_size + 1
            if node.val == sum_ // size:
                nonlocal ans
                ans += 1
            return sum_, size
        dfs(root)

        return ans